### PR TITLE
Add the connected components labeling algorithm

### PR DESCRIPTION
**Pull request by edgarriba**
### Description

This PR implements  the Connected-component labelling (CCL) algorithm in PyTorch as requested here #https://github.com/kornia/kornia/issues/866.

Colab with example: https://colab.research.google.com/drive/1q4GPsS1ZqpGoz5-eEyyahGJa6qmmbNXt?usp=sharing

**NOTE:** The implementation is sub-optimal in terms of time performance compared to OpenCV.
We keep this as first version and later we can improve with faster kernels.

Display example:

image
image

### Status
**Work in progress**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>

  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?

</details>

### CODE REVIEW

File: kornia/contrib/connected_components.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    if len(input.shape) != 4 or input.shape[-3] != 1:
+        raise ValueError(f"Input image shape must be Bx1xHxW. Got: {input.shape}")
+
+    # precomput a mask with the valid values
+    mask = (input == 1)
+
+    # allocate the output tensors for labels
+    B, _, H, W = input.shape
+    out = torch.arange(B * W * H, device=input.device, dtype=input.dtype).reshape((B, 1, H, W))
```
Comment (justanhduc) [inline kornia/contrib/connected_components.py:43]: Maybe `reshape` -> `view`?

File: kornia/contrib/connected_components.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+def connected_components(input: torch.Tensor, num_iterations: int = 100) -> torch.Tensor:
+    r"""Computes the Connected-component labelling (CCL) algorithm.
+
+    The implementation is an adaptation of the following repository:
+
+    https://gist.github.com/efirdc/5d8bd66859e574c683a504a4690ae8bc
+
+    Args:
+        input: the binarized input image with shape :math:`(B, 1, H, W)`.
```
Comment (justanhduc) [inline kornia/contrib/connected_components.py:18]: How about using the decorator `perform_keep_shape_image` to support any image dimension?

File: test/test_contrib.py [hunk 3]
```diff
@@ -7,6 +7,70 @@
 from kornia.testing import assert_close
 
 
+class TestConnectedComponents:
```
Comment (justanhduc) [inline test/test_contrib.py:11]: If it can support image of any dimension, let's add a test for shapes.

File: kornia/contrib/connected_components.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    # precomput a mask with the valid values
+    mask = (input == 1)
+
+    # allocate the output tensors for labels
+    B, _, H, W = input.shape
+    out = torch.arange(B * W * H, device=input.device, dtype=input.dtype).reshape((B, 1, H, W))
+    out[~mask] = 0
+
+    for _ in range(num_iterations):
```
Comment (justanhduc) [inline kornia/contrib/connected_components.py:46]: I'm wondering, to speed up, should we calculate the receptive field based on `num_iterations` then perform only 1 convolution?
